hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/StreamBlockInput.java

```diff
@@ -0,0 +1,777 @@
+
+  private void validateBlock(
+      ContainerProtos.ContainerCommandResponseProto response
+  ) throws IOException {
+
+    ReadBlockResponseProto readBlock = response.getReadBlock();
+    for (ReadChunkResponseProto readChunk : readBlock.getReadChunkList()) {
+      List<ByteString> byteStrings;
+
+      ContainerProtos.ChunkInfo chunkInfo =
+          readChunk.getChunkData();
+      if (chunkInfo.getLen() <= 0) {
+        throw new IOException("Failed to get chunk: chunkName == "
+            + chunkInfo.getChunkName() + "len == " + chunkInfo.getLen());
+      }
+      byteStrings = readChunk.getDataBuffers().getBuffersList();
+      long buffersLen = BufferUtils.getBuffersLen(byteStrings);
+      if (buffersLen != chunkInfo.getLen()) {
+        // Bytes read from chunk should be equal to chunk size.
+        throw new OzoneChecksumException(String.format(
+            "Inconsistent read for chunk=%s len=%d bytesRead=%d",
+            chunkInfo.getChunkName(), chunkInfo.getLen(),
+            buffersLen));
+      }
+
+
+      if (verifyChecksum) {
+        ChecksumData checksumData = ChecksumData.getFromProtoBuf(
+            chunkInfo.getChecksumData());
+        int startIndex = (int) readChunk.getChunkData().getOffset() / checksumData.getBytesPerChecksum();
+
+        // ChecksumData stores checksum for each 'numBytesPerChecksum'
+        // number of bytes in a list. Compute the index of the first
+        // checksum to match with the read data
+
+        Checksum.verifyChecksum(byteStrings, checksumData, startIndex,
+            false);
+      }
+    }
+  }
+
+  @VisibleForTesting
+  protected synchronized void setBuffers(List<ByteBuffer> buffers) {
+    this.buffers = buffers;
+  }
+
+  private boolean shouldRetryRead(IOException cause) throws IOException {
+    RetryPolicy.RetryAction retryAction;
+    try {
+      retryAction = retryPolicy.shouldRetry(cause, ++retries, 0, true);
+    } catch (IOException e) {
+      throw e;
+    } catch (Exception e) {
+      throw new IOException(e);
+    }
+    return retryAction.action == RetryPolicy.RetryAction.RetryDecision.RETRY;
+  }
+
+  @VisibleForTesting
+  public boolean isVerifyChecksum() {
+    return verifyChecksum;
+  }
+
+  private void refreshBlockInfo(IOException cause) throws IOException {
+    LOG.info("Attempting to update pipeline and block token for block {} from pipeline {}: {}",
+        blockID, pipelineRef.get().getId(), cause.getMessage());
+    if (refreshFunction != null) {
+      LOG.debug("Re-fetching pipeline and block token for block {}", blockID);
+      BlockLocationInfo blockLocationInfo = refreshFunction.apply(blockID);
+      if (blockLocationInfo == null) {
+        LOG.warn("No new block location info for block {}", blockID);
+      } else {
+        setPipeline(blockLocationInfo.getPipeline());
+        LOG.info("New pipeline for block {}: {}", blockID,
+            blockLocationInfo.getPipeline());
+
+        tokenRef.set(blockLocationInfo.getToken());
```

Comment:
Probable null value being set as token?

Reply:
Yes, it will be checked in `ContainerProtocalCalls.readBlock()`. I think we don't need worry about it.